include/boost/type_traits/intrinsics.hpp

```diff
@@ -183,25 +183,39 @@
 #   if (!defined(__GLIBCXX__) || (__GLIBCXX__ >= 20080306 && __GLIBCXX__ != 20080519)) && __has_feature(is_empty)
 #     define BOOST_IS_EMPTY(T) __is_empty(T)
 #   endif
-#   if __has_feature(has_trivial_constructor)
+#   if defined(__has_builtin) && __has_builtin(__is_trivially_constructible)
```

Comment:
This doesn't work if `__has_builtin` is not defined. The correct way to do this is:

```
#if defined(__has_builtin)
#if __has_builtin(__is_trivially_constructible)
...
#endif
#elif ...
```

The same everywhere else below.

PS: I know clang defines `__has_builtin`, that's why it compiles. But since you're checking it, the code must work correctly if it is not defined too.

Reply:
Hm, I thought this'd be OK since we're inside `#if defined(BOOST_CLANG) && ...`? It's what Boost.Config does in boost/config/compiler/clang.hpp. Happy to change it though.

Reply:
```c++
#if defined(__has_builtin)
#if __has_builtin(__is_trivially_constructible)
...
#endif
#elif ...
```
This won't work, since the `#elif` only kicks in if Clang doesn't have `__has_builtin`.  We'd need to repeat the branch or define a helper macro.

Since we're inside `BOOST_CLANG` and every version of clang (since 3.0) has `__has_builtin`, maybe it'd be OK just to remove that unnecessary check?

Reply:
> Hm, I thought this'd be OK since we're inside `#if defined(BOOST_CLANG) && ...`?

That condition is also checking for `__has_feature` being defined, so `BOOST_CLANG` is not taken as an indicator for its presence. So you shouldn't be assuming that `__has_builtin` is defined either.

> It's what Boost.Config does in boost/config/compiler/clang.hpp. Happy to change it though.

If you mean [this](https://github.com/boostorg/config/blob/8e94132dad65f171f06182e03eaf02c406552713/include/boost/config/compiler/clang.hpp#L330) then this is a bug in Boost.Config. It is fixed as part of https://github.com/boostorg/config/pull/442.

> This won't work, since the `#elif` only kicks in if Clang doesn't have `__has_builtin`. We'd need to repeat the branch or define a helper macro.

Yes, I wasn't suggesting to use exactly that preprocessor code, just illustrated how the conditions should be rearranged. You could move the check for `__has_builtin` to the `__has_feature` check above, for example, if that keeps the code correct and simple.

> Since we're inside `BOOST_CLANG` and every version of clang (since 3.0) has `__has_builtin`, maybe it'd be OK just to remove that unnecessary check?

Personally, I would not rely on that. Just do the check that you actually need instead of hoping it'll work.

Reply:
> You could move the check for `__has_builtin` to the `__has_feature` check above, for example, if that keeps the code correct and simple.

Yes, that looks like the right approach. Thanks!